internal/scheduler/upsert.go

```diff
@@ -0,0 +1,323 @@
+// with name topicName. writeMessageIds maps partition indices to the id of the most recently
+// received Pulsar message for that partition.
+func IdempotentUpsert(ctx context.Context, db *pgxpool.Pool, topicName string, writeMessageIds map[int32]pulsar.MessageID, tableName string, schema string, records []interface{}) error {
+	if len(records) == 0 {
+		return nil
+	}
+	return db.BeginTxFunc(ctx, pgx.TxOptions{
+		IsoLevel:       pgx.ReadCommitted,
+		AccessMode:     pgx.ReadWrite,
+		DeferrableMode: pgx.Deferrable,
+	}, func(tx pgx.Tx) error {
+
+		err := IdempotencyCheck(ctx, tx, topicName, writeMessageIds)
+		if err != nil {
+			return err
+		}
+
+		err = CopyProtocolUpsert(ctx, tx, tableName, schema, records)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// Upsert is like [IdempotentUpsert], except the it does no idempotency check.
+func Upsert(ctx context.Context, db *pgxpool.Pool, tableName string, schema string, records []interface{}) error {
+	if len(records) == 0 {
+		return nil
+	}
+	return db.BeginTxFunc(ctx, pgx.TxOptions{
+		IsoLevel:       pgx.ReadCommitted,
+		AccessMode:     pgx.ReadWrite,
+		DeferrableMode: pgx.Deferrable,
+	}, func(tx pgx.Tx) error {
+		return CopyProtocolUpsert(ctx, tx, tableName, schema, records)
+	})
+}
+
+func IdempotencyCheck(ctx context.Context, tx DBTX, topicName string, writeMessageIds map[int32]pulsar.MessageID) error {
+
+	// Load the message ids stored in postgres.
+	queries := New(tx)
+	sqlWriteMessageIds, err := queries.GetTopicMessageIds(ctx, topicName)
+	if err != pgx.ErrNoRows && err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Check if the data stored in postgres is more recent than what we're trying to write.
+	staleWrites := make([]StaleWrite, 0)
+	for _, sqlWriteMessageId := range sqlWriteMessageIds {
+
+		// Convert from the SQL-specific representation of message ids to one that can be used for comparison.
+		dbMessageId := pulsarMessageIdFromPulsarRecord(sqlWriteMessageId)
+
+		// If the id loaded from the database is more recent than the one provided, abort the transaction.
+		// Since the data we're trying to write is stale.
+		if writeMessageId, ok := writeMessageIds[dbMessageId.PartitionIdx()]; ok {
+			isGreater, err := dbMessageId.Greater(writeMessageId)
+			if err != nil {
+				return err
+			}
+			if isGreater {
+				staleWrites = append(staleWrites, StaleWrite{
+					WriteMessageId: writeMessageId,
+					DbMessageId:    dbMessageId,
+				})
+			}
+		}
+	}
+	if len(staleWrites) > 0 {
+		return &ErrStaleWrite{
+			Topic:       topicName,
+			StaleWrites: staleWrites,
+		}
+	}
+
+	// Otherwise, we have more recent data than what is already stored in the db that we should write.
+	// Update the message id in postgres to reflect the data to be written.
+	//
+	// TODO: Add a call to insert these in a single call.
+	for partitionIdx, writeMessageId := range writeMessageIds {
```

Comment:
Are we sure this will work?  Having read the pulsar docs a bit more, I'm not convinced that that each topic is served by exactly one consumer for a keyShared subscription.  

e.g. from [here](https://streamnative.io/en/blog/engineering/2021-08-25-scalable-stream-processing-with-pulsars-key-shared-subscription/)
```
Pulsar achieves these guarantees by hashing the incoming key values and distributing the hashes evenly across all of the consumers on the subscription. Thus, we know that messages with the same key will generate the same hash value and consequently get sent to the same consumer as the previous messages with that key.

By ensuring that all messages with the same key are sent to the same consumer, that consumer is guaranteed to receive all of the messages for a given key and in the order they were received, which matches the streaming consumption semantics. Let’s explore a real-world use case where Pulsar’s Key_Shared subscription could be used effectively
```
 and
 ```
 Pulsar’s Key_Shared subscription type allows you to have both high throughput and stateful processing on a single topic.
 ```
  This implies to me that sharing of keys between subscribers isn't limited to the simple solution of having each subscriber listen to a subset of the partitions.